flocker/node/_deploy.py

```diff
@@ -405,13 +427,22 @@ def __init__(self, hostname, volume_service, docker_client=None,
         self.network = network
         self.volume_service = volume_service
 
-    def discover_local_state(self):
+    def discover_local_state(self, cluster_state):
         """
         List all the ``Application``\ s running on this node.
 
-        :returns: A ``Deferred`` which fires with a ``NodeState``
+        :param Deployment cluster_state: The last known state of the whole
+            cluster.
+
+        :returns: A ``Deferred`` which fires with a ``NodeApplications``
             instance.
         """
+        # XXX delete manifestations code that is unrelated to
+        # applications, return NodeApplications instance.  Application's
+        # attached volumes/manifestations can be calculated based on
+        # passed-in cluster_state. This has a potential race condition
+        # unfortunately, but probably not an issue in practice.
```

Comment:
This comment is kind of unclear but I'm guessing that you plan to delete the code that calls `volume_service.list()` and replace it with code that looks up manifestations from the `cluster_state` object?  If so, sounds great.

Concerning the race condition, it seems like it's okay for this iteration of the implementation to have that problem but I think we should have a plan for fixing it later (it strikes me that this is the kind of misbehavior that we're often complaining about from Docker :smile:).

One simple (to think about but maybe hard to implement) solution seems like it would be to make sure this deployer (ie, the one for the container convergence agent) never gets asked to do anything until the dataset deployer has had a chance to discover and report (such that its state is part of the `cluster_state` passed here).  This means the this deployer (container convergence agent) will always be making decisions based on a real (if possibly slightly stale - but that doesn't matter) state instead of a likely erroneous state (ie, "there are no manifestations") which is not just state but simply wrong.  This, I think, eliminates the race condition.

The complications I can think of relating to this are:

  * Currently, the race condition goes both ways.  The dataset convergence agent depends on application state to make decisions correctly ("Is there an application running that is using this manifestation?") and vice versa ("Does a manifestation exist that matches this volume attached to this container?").  So you'd actually want both agents to get a chance to report their state to the other before either one started making decisions - but state discovery for applications depends on state discovery for datasets so bleugh it's all a big tangled mess.  The saving grace here is that eventually we'll use leases instead of having the datasets convergence agent peek into the application state which probably untangles the circularity.  So possibly this complication can be ignored in the long run.
  * Having the agents run separately (which we certainly want) makes it harder to actually implement this ordering.  I can imagine ways that the protocol might change to provide enough information to let the control service do some ordering stuff but that only helps once the agents are event driven (since right now they'll do discovery whenever they want, not just when the control service tells them to).  Also, doing this without making the agents heterogeneous and putting knowledge of them into the control service means extending the protocol even further to let agents themselves somehow declare their ordering requirements and there's more bleugh big tangled mess.

I think I had an idea for a supposedly less simple solution when I started typing this comment but now I seem to have lost it.  Maybe we should discuss this on a higher bandwidth channel and you can either explain some other solution to me or we can work one out together.